Animation research

Berthalamew opened the pull request:
- defined the animation graph definition as well as the member classes of it and some of the member functions
- update definitions for c_animation_channel and c_interpolator_control
- fixes compilation issues with constant tagblocks and string ids

- this is beneficial since it allows us to read and access animation tags incase someone wants to use them later on in project

Berthalamew committed:
add animation graph definition

- add c_animation_id class and member functions
- update c_animation_channel members

Berthalamew committed:
update animation manager class

- include server offsets for functions
- rename fields
- fix constant tagblocks not working
- add HS_ID_NONE

Berthalamew committed:
update member names in c_animation_channel

Berthalamew committed:
added member functions for c_model_animation

- updated member names in the animation definitions

Berthalamew committed:
add some member functions for c_model_animation_graph

- rename e_model_flags to e_node_model_flags
- update string_id operator overloads to work with constant string_id's

Berthalamew committed:
add member functions for c_model_animation_runtime_data

Berthalamew committed:
add member functions for c_animation_graph_resources

- add member functions for c_animation_data_sizes
- rename some member variables

Berthalamew committed:
label the rest of the c_interpolator_control class variables

- add enabled and finished member functions

Berthalamew committed:
Fix build issue

Berthalamew committed:
update c_animation_channel members

- specify bitflag members
- make c_quantized_orientation members public for the time being

Berthalamew committed:
add requested changes from himanshu

- add asserts for class sizes
- update enum fields

Berthalamew committed:
rename interpolator_control to animation_interpolation

- this is the proper module name

nukeulater commented on xlive/Blam/Common/Common.h:
```diff
@@ -42,7 +42,7 @@ static_assert (sizeof(STRUCT) == _SIZE, "Invalid size for struct (" #STRUCT")")
 
 #define TAG_BLOCK_SIZE_ASSERT(tagblock,size)\
-static_assert (sizeof(tagblock) == (size),"Invalid Size for TagBlock <" #tagblock ">")
+static_assert (sizeof(tagblock) == size,"Invalid Size for TagBlock <" #tagblock ">")
```
why?

Berthalamew replied:
Was looking at the assert while working on the definitions and the brackets seemed unnecessary.
I can add them back if you want.

nukeulater replied:
macro pitfalls: https://gcc.gnu.org/onlinedocs/gcc-3.0.1/cpp_3.html#SEC26

Berthalamew replied:
reverted back to before

nukeulater commented on xlive/Blam/Engine/animations/animation_definitions.cpp:
```diff
@@ -0,0 +1,303 @@
+
+float c_model_animation::get_authored_duration() const
+{
+	return this->frame_count * 0.0333333333f;
```
`frame_count / 30.0f` or `frame_count * (1.0f / 30.0f)` would be more preferable

Berthalamew replied:
changed to this->frame_count * (1.0f / 30.0f)

Berthalamew committed:
add requested changes

- revert macro change
- change get_authored_duration() member function to return (this->frame_count * (1.0f / 30.0f)). the compiler should optimize it to (this->frame_count * 0.0333333333f ) anyways
